sys/vm/vm_map.c

```diff
@@ -4544,8 +4704,16 @@ vm_map_remove_locked(vm_map_t map, vm_offset_t start, vm_offset_t end)
 	}
 
 	result = vm_map_delete(map, start, end, true);
-	if (vm_map_reservation_is_unmapped(map, reservation))
-		vm_map_reservation_delete_locked(map, reservation);
+	if (vm_map_reservation_is_unmapped(map, reservation)) {
+#ifdef CHERI_CAPREVOKE
+		if (quarantine_unmapped_reservations &&
+		    start < VM_MAXUSER_ADDRESS) {
+			vm_map_lookup_entry(map, start, &entry);
+			vm_map_entry_quarantine(map, entry);
```

Comment:
I'd assert that this entry covers the entire range `[start, end)`.

Also, is it possible that userspace can quarantine the same range multiple times? I don't think `vm_map_entry_quarantine()` handles that, it assumes that the entry doesn't already belong to the RB tree.

Reply:
I'll add an assert and a comment.

Portions of reservations make a one-way transition to unmapped.  Once the entire range covered by the reservation is unmapped, we transition to quarantine (also one-way).  In the current branch that's the only way quarantine happens so it can only happen once. An eventual merge to the coexecve branch, we'll also quarantine memory owned by other processes on fork, but that should also only happen once.